sklearn/utils/tests/test_fixes.py

```diff
@@ -28,3 +31,25 @@ def test_expit():
 
 def test_divide():
     assert_equal(divide(.6, 1), .600000000000)
+
+
+def test_astype_copy_memory():
+    a_int32 = np.ones(3, np.int32)
+
+    # Check that dtype conversion works
+    b_float32 = astype(a_int32, dtype=np.float32, copy=False)
+    assert_equal(b_float32.dtype, np.float32)
+
+    # Changing dtype forces a copy even if copy=False
+    assert_false(np.may_share_memory(b_float32, a_int32))
+
+    # Check that copy can be skipped if requested dtype match
+    c_int32 = astype(a_int32, dtype=np.int32, copy=False)
+    assert_true(np.may_share_memory(c_int32, a_int32))
```

Comment:
By reading the docstring, it should never yield false negatives. It can yield false positives but not for contiguous arrays, only for complex views / slices IIRC.